In [55]:
import numpy as np
import pandas as pd

# QUESTION 1

### PART A

In [56]:
# The file has already been downloaded to my computer so I am calling it directly from here
# Create a DataFrame type object called shopify to store all the data inside this row
shopify = pd.read_csv('Shopify.csv')

In [57]:
# To obatin a quick statistical overview of this dataset
shopify.describe()

order_id      shop_id      user_id   order_amount  total_items
count  5000.000000  5000.000000  5000.000000    5000.000000   5000.00000
mean   2500.500000    50.078800   849.092400    3145.128000      8.78720
std    1443.520003    29.006118    87.798982   41282.539349    116.32032
min       1.000000     1.000000   607.000000      90.000000      1.00000
25%    1250.750000    24.000000   775.000000     163.000000      1.00000
50%    2500.500000    50.000000   849.000000     284.000000      2.00000
75%    3750.250000    75.000000   925.000000     390.000000      3.00000
max    5000.000000   100.000000   999.000000  704000.000000   2000.00000

A quick look will tell us that the maximum value of order_amount column is completely out of bound comparing to the other data in this column. This outlier greatly affect the mean value of the order_amount column, hence, greatly push the AOV value up to $3145

In [58]:
print('Wrong value of AOV is the mean of order_amount: $', df.order_amount.mean())

Wrong value of AOV is the mean of order_amount: $ 3145.128


There are a few ways to improve this data:

Option 1) We get rid of all outliers by picking value only within the interquartile range (IQR)

Option 2) We can choose to pick the median of this data set

## Option 1 - IQR

### PART B

We will need to filter this data set to only take in data that has entries in the 'order_amount' to be within the 1st and 3rd quartile. The metric to report after this filter step is still going to be the MEAN.

In [59]:
Q1 = shopify['order_amount'].quantile(0.25) # 1st quartile of the data
Q3 = shopify['order_amount'].quantile(0.75) # 3rd quartile of the data
IQR = Q3 - Q1 

# range of value where we accept data
lower_range = Q1 - 1.5 * IQR
upper_range = Q3 + 1.5 * IQR

# Filtering Values between Q1-1.5IQR and Q3+1.5IQR 
shopify_filtered = (shopify.order_amount > lower_range) & (shopify.order_amount < upper_range) # Boolean 
shopify_filtered = shopify[shopify_filtered] # Slice only the true value that falls within this range

In [60]:
# Now, with the outliers filtered, we obtain a better data set
shopify_filtered.describe()

order_id      shop_id      user_id  order_amount  total_items
count  4859.000000  4859.000000  4859.000000   4859.000000  4859.000000
mean   2497.395966    49.852645   849.905742    293.715374     1.950196
std    1443.356555    29.049171    86.887496    144.453395     0.919791
min       1.000000     1.000000   700.000000     90.000000     1.000000
25%    1244.500000    24.000000   776.000000    162.000000     1.000000
50%    2498.000000    50.000000   850.000000    280.000000     2.000000
75%    3749.500000    74.000000   925.000000    380.000000     3.000000
max    5000.000000   100.000000   999.000000    730.000000     5.000000

### PART C

In [61]:
print('The AOV of this filtered data set is: $', shopify_filtered['order_amount'].mean())

The AOV of this filtered data set is: $ 293.7153735336489


## Option 2: Median

### PART B

The metric to report here is the median. The median is less affected by outliers. For a smaller data set where we can not afford to get rid of data, this method should be more preferable.

In [62]:
print('The AOV of the original data set is: $', shopify['order_amount'].median())

The AOV of the original data set is: $ 284.0


# QUESTION 2

### PART A

The Shippers and Orders table are joined by ShipperID. From here, the total number all 'Speedy Express' shipping is counted and printed out. 

### PART B

The Orders and Employees table are joined by EmployeeID. Employees' Last Name and its count is tabulated into a new list. The last name with the most count is called out of this list.

In [63]:
### PART C

The Orders, OrderDetails, Product and Customers are joined together by OrderID, ProductID and CustomerID. Product Name and the total number of order associated to it are tabulated into a new list. Note that only the entries with Country is Germany are kept. The product with the most count is called out of this list.  